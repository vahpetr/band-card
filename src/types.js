export const ValueType = {
    YIELD: "Yield",
    SPREAD: "Spread",
    PRICE: "Price"
}