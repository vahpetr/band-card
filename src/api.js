// @ts-check

import { ValueType } from "./types";

/**
 * Fetch bond data
 */
export const fetchBondData = filter =>
  fakeFetchBondData("//fakeapi.bond.com", filter);

/**
 * Fake fetch bond data
 */
const fakeFetchBondData = (_, filter) =>
  new Promise((resolve, reject) => {
    window.setTimeout(() => {
      try {
        const data = generateBondData(filter);
        resolve({ data });
      } catch (ex) {
        reject(ex.message);
      }
    }, 0);
  });

/**
 * Generate number from range
 * @param {number} min
 * @param {number} max
 */
const rnd = (min, max) => Math.random() * (max - min) + min;

/**
 * Generate bond data array
 */
const generateBondData = filter => {
  const fillYield = (_, index) => rnd(20, 30) + Math.log(index);
  const fillSpread = (_, index) => rnd(2, 4) + index / 30;
  const fillPrice = (_, index) => rnd(60, 80) + index;

  const getMapper = valueType => {
    switch (valueType) {
      case ValueType.YIELD:
        return fillYield;
      case ValueType.SPREAD:
        return fillSpread;
      case ValueType.PRICE:
        return fillPrice;
      default:
        throw new Error(`Bound query not supported ${valueType} value type.`);
    }
  };

  const mapper = getMapper(filter.valueType);
  const data = Array(70 * 3)
    .fill(undefined)
    .map(mapper);

  return data;
};
