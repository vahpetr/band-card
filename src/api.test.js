import { ValueType } from "./types";
import { fetchBondData } from "./api";

jasmine.DEFAULT_TIMEOUT_INTERVAL = 1000;

describe("Fetch bond data api", () => {
  it("should return result", async () => {
    const result = await fetchBondData({
      valueType: ValueType.SPREAD
    });
    expect(result.data.length).toBe(210);
  });

  it("should be throw exception", async () => {
    expect.assertions(1);
    const task = fetchBondData({
      valueType: "fake"
    });
    await expect(task).rejects.toThrowErrorMatchingSnapshot();
  });
});
