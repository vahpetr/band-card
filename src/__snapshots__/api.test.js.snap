// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch bond data api should be throw exception 1`] = `"received is not a function"`;
